exchange.go

```diff
@@ -53,79 +53,123 @@ func (re *RabbitExchangeImpl) Close() error {
 }
 
 func (re *RabbitExchangeImpl) SendTo(name, exchangeType string, durable, autoDelete bool, key string) MessageHandleFunc {
-	return func(ctx context.Context, message []byte) error {
-		if ctx == nil {
-			return ErrNilContext
-		}
 
-		if ctx.Err() != nil {
-			return ctx.Err()
-		}
+	var conn *amqp.Connection
+	var ch *amqp.Channel
+	var version int64 = 0
 
-		conn, err := re.getEventConnection()
-		if err != nil {
-			return err
-		}
+	var connM sync.RWMutex
+	var chM sync.RWMutex
 
-		if ctx.Err() != nil {
-			return ctx.Err()
-		}
-
-		var firstError error
-		var ch *amqp.Channel
-		var try int
-		for try = 1; try <= 3; try++ {
-			if ctx.Err() != nil {
-				return ctx.Err()
+	getConnection := func() (*amqp.Connection, error) {
+		currectConn := func() *amqp.Connection {
```

Comment:
correct? current?

Reply:
Are you doing this in a function just so that you can defer? Maybe it's just me but it made everything a bit harder to read.

Reply:
Yes, it is in a function for the defer. 

And trust me, the locking/unlocking without a defer is much worse.